astore/server/main.go

```diff
@@ -135,13 +136,19 @@ func Start(targetURL, cookieDomain, oAuthType string, astoreFlags *astore.Flags,
 		return fmt.Errorf("could not initialize auth server - %s", err)
 	}
 
-	authWeb, err := oauth.New(rng, oauth.WithFlags(oauthFlags), auth.FetchCredentialOpts(oAuthType))
+	reqAuth, err := oauth.New(rng, oauth.WithFlags(oauthFlags), auth.FetchCredentialOpts(oAuthType))
+	if err != nil {
+		return fmt.Errorf("could not initialize oauth authenticator - %s", err)
+	}
+	optAuth, err := oauth.New(rng, oauth.WithFlags(optAuthFlags), auth.FetchCredentialOpts(oAuthType))
```

Comment:
Ok, one problem here. The end goal is for the user to authenticate once with our systems, and be good for the rest of the day.

At the end of the login process, 2 things happen:

a) An authentication token is returned via gRPC to the CLI interface, for use by enkit.

b) A cookie is set in the browser, containing an encrypted token. For use in web interfaces.

Web interfaces that care about b) are the enkit proxy (so user does not have to authenticate separately with it), and b) some of the astore URLs (I believe authenticated users can download stuff via browser for some artifacts - if not implemented yet, I'd sure like it). Pretty much, anything we put on GAE should check the credentials cookie, just like astore or the proxy.

Now... I'm confident that the code for the CLI works properly here. I'm not confident about the code for the browser.

I suspect the last `PerformAuth` you invoke overwrites the cookie for the browser. Further, I fear that if the user only half completes the authentication (example: only authenticates with google, and not github), the user will still get a valid authentication cookie!

Two easy alternative:
1) The oauth library supports a different cookie name per instance of the object. We could use it? I would prefer solution 2, for safety.

2) In the oauth library, break the `PerformAuth` method in two: 1) That checks and extracts the state, 2) That issues an authentication cookie. To not break other users of the API, `PerformAuth` is then changed to simply invoke function 1 and function 2 in order.

From MultiOauth, then, you only call function 1 from the oauth providers. And call function 2 only at the end of the process (eg, when all authentications have completed).

Reply:
The solution for this really doesn't require any change in code yeah?

Just add `--opt-base-cookie=external` or something